### PR TITLE
WIP: review and refactoring of @Qualified PR

### PR DESCRIPTION
**Pull request by stevenschlansker**
WIP on some thoughts on #1597

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -10,6 +10,8 @@
   - KotlinSqlObjectPlugin will install forgotten SqlObjectPlugin for you
   - ClasspathSqlLocator allows disabling comment stripping and deprecate static API
   - KotlinMapper respects `@PropagateNull`
+  - Freemarker allows customizing Configuration
+  - FreemarkerSqlLocator should now be created via FreemarkerConfig so it shares config
```
Comment (qualidafial) [inline RELEASE_NOTES.md]: Part of a different PR?

File: core/src/main/java/org/jdbi/v3/core/mapper/NVarcharMapper.java [hunk 2]
```diff
@@ -13,19 +13,23 @@
...
 import org.jdbi.v3.core.qualifier.NVarchar;
-import org.jdbi.v3.core.statement.StatementContext;
+import org.jdbi.v3.core.qualifier.QualifiedType;
 
 /**
  * Column mapper for Java {@code @NVarchar String} qualified type.
  */
-@NVarchar
-class NVarcharMapper implements ColumnMapper<String> {
+class NVarcharMapper implements QualifiedColumnMapperFactory {
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/mapper/NVarcharMapper.java]: Should rename to `NVarcharMapperFactory`

File: core/src/main/java/org/jdbi/v3/core/qualifier/Qualifiers.java [hunk 3]
```diff
@@ -64,17 +57,16 @@ public void setRegistry(ConfigRegistry registry) {
      * which searches for annotations themselves annotated with {@link Qualifier}.
      *
      * Resolvers should return static results that don't vary over time; therefore their results are cached.
-     * Adding a resolver will clear the local qualifier resolution cache, but is currently not guaranteed to have any effect on {@code SqlObjects}.
-     * Add any resolvers you may need <em>before</em> using any qualifiers-related features!
+     * Add any resolvers you may need <em>before</em> using any types that may be qualified - cached qualifiers
+     * are not guaranteed to update!
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/qualifier/Qualifiers.java]: Javadoc needs to be updated on this method

File: core/src/main/java/org/jdbi/v3/core/qualifier/Qualifiers.java [hunk 4]
```diff
@@ -64,17 +57,16 @@ public void setRegistry(ConfigRegistry registry) {
...
      */
-    public Qualifiers addResolver(Function<AnnotatedElement, Set<Annotation>> resolver) {
-        resolvers.add(resolver);
-        if (registry != null) {
-            QUALIFIER_CACHE.clear(registry);
-        }
+    @SafeVarargs
+    public final Qualifiers registerQualifier(Class<? extends Annotation>... annotations) {
+        registered.addAll(Arrays.asList(annotations));
+        QUALIFIER_CACHE.clear(registry);
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/qualifier/Qualifiers.java]: A local config change (e.g. on a SQL Object method) invalidates the global qualifier cache? This means that qualifiers could get loaded and reloaded and get different values, depending on the last time a SQL object registered a qualifier.

It's even worse if you have two different SQL objects registering different qualifiers.

If the qualifier cache is global, then the qualifier registry needs to be global too.

File: freemarker/src/main/java/org/jdbi/v3/freemarker/FreemarkerConfig.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ */
+package org.jdbi.v3.freemarker;
+
+import java.util.Optional;
+
+import freemarker.cache.ClassTemplateLoader;
+import freemarker.template.Configuration;
+import org.jdbi.v3.core.config.JdbiConfig;
+
+public class FreemarkerConfig implements JdbiConfig<FreemarkerConfig> {
```
Comment (qualidafial) [inline freemarker/src/main/java/org/jdbi/v3/freemarker/FreemarkerConfig.java]: unrelated to qualifiers

File: freemarker/src/main/java/org/jdbi/v3/freemarker/FreemarkerEngine.java [hunk 6]
```diff
@@ -28,13 +28,13 @@
     @Override
     public String render(String sql, StatementContext ctx) {
         try {
-            Template template = new Template(null, sql, null);
+            Template template = new Template(null, sql,
+                    ctx.getConfig(FreemarkerConfig.class).getFreemarkerConfiguration());
```
Comment (qualidafial) [inline freemarker/src/main/java/org/jdbi/v3/freemarker/FreemarkerEngine.java]: Unrelated to qualifiers

File: freemarker/src/main/java/org/jdbi/v3/freemarker/FreemarkerSqlLocator.java [hunk 7]
```diff
@@ -24,9 +24,17 @@
  * Locates SQL in {@code .sql.ftl} Freemarker files on the classpath.
  */
 public class FreemarkerSqlLocator {
+    /**
+     * @deprecated don't use static scope
+     */
+    @Deprecated
```
Comment (qualidafial) [inline freemarker/src/main/java/org/jdbi/v3/freemarker/FreemarkerSqlLocator.java]: ditto

File: freemarker/src/main/java/org/jdbi/v3/freemarker/internal/UseFreemarkerSqlLocatorImpl.java [hunk 8]
```diff
@@ -33,13 +34,12 @@
 public class UseFreemarkerSqlLocatorImpl implements Configurer {
     @Override
     public void configureForType(ConfigRegistry registry, Annotation annotation, Class<?> sqlObjectType) {
-        SqlLocator locator = (type, method, config) -> {
-            String templateName = SqlAnnotations.getAnnotationValue(method).orElseGet(method::getName);
-            findTemplate(sqlObjectType, templateName);
-            return templateName;
-        };
+        SqlLocator locator = (type, method, config) ->
```
Comment (qualidafial) [inline freemarker/src/main/java/org/jdbi/v3/freemarker/internal/UseFreemarkerSqlLocatorImpl.java]: ditto

File: RELEASE_NOTES.md (hunk 1 above)
Comment (stevenschlansker) [reply RELEASE_NOTES.md]: I did a `git merge master` into this branch to resolve the conflicts, as I didn't want to rebase across Marnes' work

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/HandlerFactory.java [hunk 9]
```diff
@@ -16,13 +16,15 @@
...
+import org.jdbi.v3.core.config.internal.Injected;
+
 /**
  * Creates Handler objects for methods annotated with a specific SQL method annotation, which satisfy the contract of
  * that annotation.
  *
  * @see SqlOperation
  */
-public interface HandlerFactory {
+public interface HandlerFactory extends Injected {
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/HandlerFactory.java]: HandlerFactory is stateless--you are adding config state to objects that get reused between many threads. This wants to be a parameter to `buildHandler`.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/Handlers.java [hunk 10]
```diff
@@ -43,12 +45,21 @@ private Handlers(Handlers that) {
         factories.addAll(that.factories);
     }
 
+    @Override
+    public void setRegistry(ConfigRegistry registry) {
+        this.registry = registry;
+        factories.forEach(hf -> hf.setRegistry(registry));
```
Comment (qualidafial) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/Handlers.java]: Race condition. Multiple threads could be inspecting a new SQL object at once, and they would step on each other with their own config

**Comment by hgschmie** [comment]
same as #1597

**Pull request closed without merging**